fixes (possibly?) for AudioControl mapping

joshpar opened the pull request:
fixes #2839

telephon commented on include/plugin_interface/SC_Graph.h:
```diff
@@ -35,6 +35,7 @@ struct Graph
 	uint32 mNumControls;
 	float *mControls;
 	float **mMapControls;
+    uint32 mAudioBusOffset;
```
formatting? maybe convert tabs to space?

joshpar commented:
So - I have that setting already there, and it appears that my change is using 4 spaces instead of tabs. When I reformat the file, all the other lines are changed (and when I use show invisibles in Xcode, it looks like everything else is tabbed). So - I don’t feel live the white space of stuff I didn’t touch is appropriate to this issue, but I can still update the rest of the file if you think we should. Let me know what you want.
Also - did you run the patched version to make sure it performed as expected?
Best,
Josh

> On Jul 29, 2017, at 1:26 AM, Julian Rohrhuber <notifications@github.com> wrote:
> 
> @telephon commented on this pull request.
> 
> looks good - I'll test later. Could you fix the formatting please? You may have to set the preferences in xcode (if you are using that) to use 4 spaces instead of tabs, and then reformat.
> 
> In include/plugin_interface/SC_Graph.h <https://github.com/supercollider/supercollider/pull/3063#discussion_r130216565>:
> 
> > @@ -35,6 +35,7 @@ struct Graph
>  	uint32 mNumControls;
>  	float *mControls;
>  	float **mMapControls;
> +    uint32 mAudioBusOffset;
> formatting? maybe convert tabs to space?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/supercollider/supercollider/pull/3063#pullrequestreview-53072303>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AADTrLqobdidMn2Fdi0RAHCX37VQU4Zhks5sSuyYgaJpZM4OnMQZ>.
>

telephon commented:
I'm writing a test to make it a little less dependent on setup.

telephon commented:
OK, done.

joshpar commented:
Thanks! Glad the fix seems to work. If other cases come up, please let me know. 
Josh

/*
Josh Parmenter
www.realizedsound.net/josh
*/

> On Jul 30, 2017, at 07:39, Julian Rohrhuber <notifications@github.com> wrote:
> 
> OK, done.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

telephon commented:
@joshpar wrote (#2839):
>I think this can be closed. But I'll leave some notes here. Unlike In and other bus UGens, AudioControl does NOT look right at the bus. It looks at a chunk of memory in the Graph that is allocated to store audio mapped data (similar to the way Controls work). As a result, when the Graph has Audio being mapped to it, a reference to the audio bus that is mapped needs to be stored - I added a place in the Graph to store the index of the mapped audio control. This is then checked against to see if data has been written to it (touched) or not. However, I'm not positive this will work well with multiple audio mapped controls. If someone can test for that in their usual usage, and if there is a problem create an new issue, I can dig deeper for other use cases.

telephon commented:
@joshpar 

rereading the example above, I've found that the new version doesn't output any glitches, but also doesn't output any audio. 

Here is a slightly easier to use version of the first one by @jamshark70:

```
(
SynthDef(\fxbus, { |out, in|
	var sig = In.ar(in, 2);
	ReplaceOut.ar(out, HPF.ar(sig, 800));
}).add;

SynthDef(\fx, { |out, a_in = #[0, 0]|
	ReplaceOut.ar(out, HPF.ar(a_in, 800));
}).add;

SynthDef(\boop, { |out, freq = 440, gate = 1, amp = 0.1|
	var sig = Mix(VarSaw.ar(freq * [1, 1.004])) * amp,
	eg = EnvGen.ar(Env.adsr(0.01, 0.1, 0.5, 0.001, curve: 0), gate, doneAction: 2);
	sig = LPF.ar(sig, 3000);
	Out.ar(out, (sig * eg).dup(2));
}).add;
)

s.boot;
s.scope;

(
b = Bus.audio(s, 2);

p = Pbind(
	\instrument, \boop,
	\degree, Pwhite(-7, 7, inf),
	\dur, 1,
	\legato, 0.25,
	\out, b
).play;
)

f = Synth.tail(s.defaultGroup, \fxbus, [in: b, out: 0]);  // OK

f.free;

(
s.makeBundle(nil, {
	f = Synth.tail(s.defaultGroup, \fx, [\out: 0]);
	//s.sendMsg("/n_mapan", f.nodeID, \a_in, b.index, 2);
	f.map(\a_in, b);
});
)

p.stop;
f.free;
```

for me, the last bit outputs silence now (it used to outputs glitches).

telephon commented:
Also, this one crashes the server:

```


(
SynthDef(\fx, { |out|
	ReplaceOut.ar(out, HPF.ar(\in.ar(0 ! 2), 800));
}).add;
)


(
b = Bus.audio(s, 2);

s.makeBundle(nil, {
	f = Synth.tail(s.defaultGroup, \fx, [\out: 0]);
	f.map(\in, b);
});
)
```

joshpar commented:
Thanks for the example. I’ll try to work through this, but do you think you or someone else can make up an example of the version without sound that doesn’t use Pbind? I can’t always tell what is going on there with order of execution (and one thing I saw while debugging the other day had to do with this).
Thanks - I’ll look at the crashing issue as well. That example works well for me to work on this.

> On Jul 30, 2017, at 11:48 AM, Julian Rohrhuber <notifications@github.com> wrote:
> 
> Also, this one crashes the server:
> 
> 
> 
> (
> SynthDef(\fx, { |out|
> 	ReplaceOut.ar(out, HPF.ar(\in.ar(0 ! 2), 800));
> }).add;
> )
> 
> 
> (
> b = Bus.audio(s, 2);
> 
> s.makeBundle(nil, {
> 	f = Synth.tail(s.defaultGroup, \fx, [\out: 0]);
> 	f.map(\in, b);
> });
> )
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/supercollider/supercollider/pull/3063#issuecomment-318921294>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AADTrCE_wSSpKz_fgZbZElalhbcY5Nx0ks5sTNAEgaJpZM4OnMQZ>.
>

telephon commented:
Sure, here you are:

```
(
SynthDef(\fxbus, { |out, in|
	var sig = In.ar(in, 2);
	ReplaceOut.ar(out, HPF.ar(sig, 800));
}).add;

SynthDef(\fx, { |out, a_in = #[0, 0]|
	ReplaceOut.ar(out, HPF.ar(a_in, 800));
}).add;

)

s.boot;
s.scope;

(
b = Bus.audio(s, 2);

{ Out.ar(b, SinOsc.ar(330) * 0.1 ! 2) }.play;
)

f = Synth.tail(s.defaultGroup, \fxbus, [in: b, out: 0]);  // OK

f.free;

(
s.makeBundle(nil, {
	f = Synth.tail(s.defaultGroup, \fx, [\out: 0]);
	//s.sendMsg("/n_mapan", f.nodeID, \a_in, b.index, 2);
	f.map(\a_in, b);
});
)

p.stop;
f.free;
```

joshpar commented:
SO - I have it working for your solution below, and the numbers make sense to me in the code.
But - running James’s example (copied below) it doesn’t. I think the Order of Execution is messed up though. The node plot shows this:


NODE TREE Group 0
   1 group
      1070 temp__3
      1071 group
         1077 temp__0out-353201172_0
      1076 group
         1084 boop
      1073 group
         1074 system_link_audio_1
         1075 system_link_audio_1

Shouldn't node 1077 be after group 1076? I’m thinking this is why the example below now outputs silence with me new changes. Please take a look at the node tree and tell me if I’m mistaken. I’ll push a change to the remote branch after I look at the crash.

(
SynthDef(\boop, { |out, freq = 440, gate = 1, amp = 0.1|
	var sig = Mix(VarSaw.ar(freq * [1, 1.004])) * amp,
	eg = EnvGen.ar(Env.adsr(0.01, 0.1, 0.5, 0.001, curve: 0), gate, doneAction: 2);
	sig = LPF.ar(sig, 3000);
	Out.ar(out, (sig * eg).dup);
}).add;
)

(
p = ProxySpace.new.push;
~out = { \in.ar(0!2) }; ~out.play(vol: 0.2);

~player = Pbind(
	\instrument, \boop,
	\degree, Pwhite(-7, 7, inf),
	\dur, 1, \sustain, 0.2
);

~player <>> ~out;
)

p.clear; p.pop;

> On Jul 30, 2017, at 2:15 PM, Julian Rohrhuber <notifications@github.com> wrote:
> 
> Sure, here you are:
> 
> (
> SynthDef(\fxbus, { |out, in|
> 	var sig = In.ar(in, 2);
> 	ReplaceOut.ar(out, HPF.ar(sig, 800));
> }).add;
> 
> SynthDef(\fx, { |out, a_in = #[0, 0]|
> 	ReplaceOut.ar(out, HPF.ar(a_in, 800));
> }).add;
> 
> )
> 
> s.boot;
> s.scope;
> 
> (
> b = Bus.audio(s, 2);
> 
> { Out.ar(b, SinOsc.ar(330) * 0.1 ! 2) }.play;
> )
> 
> f = Synth.tail(s.defaultGroup, \fxbus, [in: b, out: 0]);  // OK
> 
> f.free;
> 
> (
> s.makeBundle(nil, {
> 	f = Synth.tail(s.defaultGroup, \fx, [\out: 0]);
> 	//s.sendMsg("/n_mapan", f.nodeID, \a_in, b.index, 2);
> 	f.map(\a_in, b);
> });
> )
> 
> p.stop;
> f.free;
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/supercollider/supercollider/pull/3063#issuecomment-318930206>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AADTrCutwKeKZaic2VmRlUqdbdKwZWhhks5sTPJrgaJpZM4OnMQZ>.
>

joshpar commented:
new fixes are in for checking. @telephon - was the sample that crashed something that you would expect to hear audio with?

telephon commented:
@joshpar no, I didn't expect audio to come out of that example. It was just there to crash the server.

joshpar commented:
I pushed some new things to that branch - and some other files had to be touched. Does the pull request automatically get updated or do I need to do a new one?
Ready for checks I think.
Josh

/*
Josh Parmenter
www.realizedsound.net/josh
*/

> On Jul 30, 2017, at 22:55, Julian Rohrhuber <notifications@github.com> wrote:
> 
> @joshpar no, I didn't expect audio to come out of that example. It was just there to crash the server.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

brianlheim commented:
> Does the pull request automatically get updated or do I need to do a new one?

If you're pushing to the remote branch that you used to make the PR, the PR gets updated automatically.

joshpar commented:
Great - that’s what I thought. Should be good for review then!


/*
Josh Parmenter
www.realizedsound.net/josh
*/

> On Jul 31, 2017, at 06:09, Brian Heim <notifications@github.com> wrote:
> 
> Does the pull request automatically get updated or do I need to do a new one?
> 
> If you're pushing to the remote branch that you used to make the PR, the PR gets updated automatically.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

telephon commented:
Excellent. This works for the following test, and doesn't crash any of the above.

```
(
SynthDef(\fxbus, { |out, in|
	var sig = In.ar(in, 2);
	ReplaceOut.ar(out, HPF.ar(sig, 800));
}).add;

SynthDef(\fx, { |out|
	ReplaceOut.ar(out, HPF.ar(\in.ar(0 ! 2), 800));
}).add;
SynthDef(\source, { |out|
	Out.ar(out, SinOsc.ar(330) * 0.1 ! 2 * Line.kr(1, 0, 0.8, doneAction:2))
}).add;
)

s.boot;
s.scope;

(
b = Bus.audio(s, 2);

fork { 
	loop { 
		Synth(\source, [\out, b]);
		1.5.wait 
	} 
};
)

f = Synth.tail(s.defaultGroup, \fxbus, [in: b, out: 0]);  // OK

f.free;

(
s.makeBundle(nil, {
	f = Synth.tail(s.defaultGroup, \fx, [\out: 0]);
	f.map(\in, b);
});
)

p.stop;
f.free;

```

telephon commented:
OK, so one more thing. So far, the audio rate controls worked as analogies to `InFeedback`, so that it would grab the previous buffer content if necessary. This allowed us to ignore node order if there is only one writer on that bus.

This is the reason why the ProxySpace example still doesn't work:

```

( 
SynthDef(\boop, { |out, freq = 440, gate = 1, amp = 0.1| 
	var sig = Mix(VarSaw.ar(freq * [1, 1.004])) * amp, 
	eg = EnvGen.ar(Env.adsr(0.01, 0.1, 0.5, 0.001, curve: 0), gate, doneAction: 2); 
	sig = LPF.ar(sig, 3000); 
	Out.ar(out, (sig * eg).dup); 
}).add
) 

(
p = ProxySpace.new.push; 
~out = { \in.ar(0 ! 2) };
~out.play(vol: 0.2);
~player = Pbind( \instrument, \boop, \degree, Pwhite(-7, 7, inf), \dur, 1, \sustain, 0.8); 
~player <>> ~out; 

)
```

While this one works:

```
(
p = ProxySpace.new.push; 
~player = Pbind( \instrument, \boop, \degree, Pwhite(-7, 7, inf), \dur, 1, \sustain, 0.8); 
~player <>> ~out; 
~out = { \in.ar(0 ! 2) };
~out.play(vol: 0.2); 
)
```

I'll post a simpler test case.

telephon commented:
Here is a test case:

```

(
SynthDef(\fxbus, { |out, in|
	var sig = InFeedback.ar(in, 2);
	ReplaceOut.ar(out, HPF.ar(sig, 800));
}).add;

SynthDef(\fx, { |out|
	ReplaceOut.ar(out, HPF.ar(\in.ar(0 ! 2), 800));
}).add;
SynthDef(\source, { |out|
	Out.ar(out, SinOsc.ar(330) * 0.1 ! 2 * Line.kr(1, 0, 0.8, doneAction:2))
}).add;
)

s.boot;
s.scope;

(
b = Bus.audio(s, 2);

fork { 
	loop { 
		Synth.tail(s.defaultGroup, \source, [\out, b]);
		1.5.wait 
	} 
};
)

f = Synth.head(s.defaultGroup, \fxbus, [in: b, out: 0]);  // OK

f.free;

// add to head: no sound.
(
s.makeBundle(nil, {
	f = Synth.head(s.defaultGroup, \fx, [\out: 0]);
	f.map(\in, b);
});
)

// add after: yes sound.
(
s.makeBundle(nil, {
	f = Synth.after(s.defaultGroup, \fx, [\out: 0]);
	f.map(\in, b);
});
)

p.stop;
f.free;

```

joshpar commented:
well - that worked because of the bug that we’re squishing. To get that back, we could possibly check if the bus was touched in the current OR last cycle, but that is really dodgy behavior in my opinion. More than happy to try it out, but in the end, we will always end up with a glitch of at least one control period at the end where a bus is no longer touched, but it WAS touched on the last cycle and that data gets repeated.
Josh

> On Jul 31, 2017, at 9:10 AM, Julian Rohrhuber <notifications@github.com> wrote:
> 
> Here is a test case:
> 
> 
> (
> SynthDef(\fxbus, { |out, in|
> 	var sig = InFeedback.ar(in, 2);
> 	ReplaceOut.ar(out, HPF.ar(sig, 800));
> }).add;
> 
> SynthDef(\fx, { |out|
> 	ReplaceOut.ar(out, HPF.ar(\in.ar(0 ! 2), 800));
> }).add;
> SynthDef(\source, { |out|
> 	Out.ar(out, SinOsc.ar(330) * 0.1 ! 2 * Line.kr(1, 0, 0.8, doneAction:2))
> }).add;
> )
> 
> s.boot;
> s.scope;
> 
> (
> b = Bus.audio(s, 2);
> 
> fork { 
> 	loop { 
> 		Synth.tail(s.defaultGroup, \source, [\out, b]);
> 		1.5.wait 
> 	} 
> };
> )
> 
> f = Synth.head(s.defaultGroup, \fxbus, [in: b, out: 0]);  // OK
> 
> f.free;
> 
> // add to head: no sound.
> (
> s.makeBundle(nil, {
> 	f = Synth.head(s.defaultGroup, \fx, [\out: 0]);
> 	f.map(\in, b);
> });
> )
> 
> // add after: yes sound.
> (
> s.makeBundle(nil, {
> 	f = Synth.after(s.defaultGroup, \fx, [\out: 0]);
> 	f.map(\in, b);
> });
> )
> 
> p.stop;
> f.free;
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/supercollider/supercollider/pull/3063#issuecomment-319117526>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AADTrJNfUEE9cnngVxqFRB7ngzSqJXl1ks5sTfx1gaJpZM4OnMQZ>.
>

telephon commented:
Modulating the bus argument of an `InFeedback` doesn't cause glitches:

```

(
SynthDef(\fxbus, { |out, in|
	var sig = InFeedback.ar(in, 2);
	ReplaceOut.ar(out, HPF.ar(sig, 800));
}).add;
SynthDef(\source, { |out|
	Out.ar(out, SinOsc.ar(330) * 0.1 ! 2)
}).add;
)


(
b = Bus.audio(s, 2);
Synth(\source, [\out, b]);

f = Synth.head(s.defaultGroup, \fxbus, [in: b, out: 0]);  // OK

fork { loop { 0.5.wait; f.set(\in, 99); 0.5.wait; f.set(\in, b); } };
)
```

Why must the `AudioControl`be different? (sorry, just asking.)

telephon commented:
https://github.com/supercollider/supercollider/blob/master/server/plugins/IOUGens.cpp#L666

joshpar commented:
InFeedback reads from a bus and takes a bus input. AudioControl maps to a bus, but the AudioControl UGen doesn't know what bus that is. It is all handled by the Graph. Like Control, AudioControl just has pointers that it deals with and no real knowledge of Audio bus. The changes I made gave it some of that, but there is no interface for AudioControl that a user accesses directly (otherwise, I would say we can add a flag that tells it whether or not to treat it like In.ar or InFeedback.ar). The 'touched' value need to match with the current buffer cycle for AudioControl to output values now (which fixed the stale bug)... but before, it was basically just stale data that was always being read.

joshpar commented:
This line:

float fbusChannel = ZIN0(0);

is what AudioControl doesn't have.

telephon commented:
In the moment the Graph is modified as a result of the server receiving a `/n_mapa` message, the first argument of that message is a bus number. When the server receives the message, it can look up the bus, can't it?

If somehow possible, it would be important to keep the feedback-like behaviour, because changing that will break lost of code, after it having been around for so long.

telephon commented:
In SC_Graph.cpp:

```
void Graph_MapAudioControl(Graph* inGraph, uint32 inIndex, uint32 inBus)
{
	if (inIndex >= GRAPHDEF(inGraph)->mNumControls) return;
	World *world = inGraph->mNode.mWorld;
	/* what is the below doing??? it is unmapping by looking for negative ints */
	if (inBus >= 0x80000000) {
		inGraph->mControlRates[inIndex] = 0;
		inGraph->mMapControls[inIndex] = inGraph->mControls + inIndex;
	} else if (inBus < world->mNumAudioBusChannels) {
	inGraph->mControlRates[inIndex] = 2;
        inGraph->mAudioBusOffsets[inIndex] = inBus;
		inGraph->mMapControls[inIndex] = world->mAudioBus + (inBus * world->mBufLength);
	}
}
```

joshpar commented:
That's what I just added - and the problem is, you need to decide if we should look at the current cycle, or the last. Even if we added something to the function def:
void Graph_MapAudioControl(Graph* inGraph, uint32 inIndex, uint32 inBus)
that let us toggle that decision, we'd also need to figure out where that value is passed in. I can make it JUST look at the past, but then we had a situation where you were testing the other day and heard silence (because something earlier in the chain had already wiped it out). Unlike InFeedback, AudioControl isn't actually writing the audio data anywhere - it is just look at a pointer that may or may not have been touched in this control cycle. With the bug, if it had been, you could end up with zeroes.

jamshark70 commented:
The other solution is for JITLib to track dependencies between proxies and order the nodes correctly. It depends on InFeedback currently, which is, in a way, cheating.

telephon commented:
In programming one person's feature is another person's cheat.

But seriously:  all synth routing in JITLib has always been independent of node order. You should not have to depend on the order of nodes at all, unless you want to minimize I/O delay time.

telephon commented:
When you rearrange the node order at runtime, you can get glitches.

joshpar commented:
at the same time, reading off an Audio Bus has ALWAYS been about node order... I understand what you mean, but I'm just not sure what the best solution here is. I AM pretty certain that, if the touched check looks at the last buffer there will be cases where sound is missing, and vice versa. And logic that looks for either / or, there is a control period that will get repeated.

telephon commented:
well, `InFeedback` was specially designed to get around this issue in the first place, and `AudioControl` has supported it always (even if in a buggy way).

I am sorry: I am still missing it why `InFeedback` gets around it even when you modulate it. There must be some way, because you never get control periods repeated there. I don't see why `AudioControl` couldn't do it the same way.

telephon commented:
```
for (int i=0; i<numChannels; ++i, in += bufLength) {
		AudioBusGuard<true> guard (unit, fbusChannel + i, maxChannel);

		float *out = OUT(i);
		int diff = bufCounter - touched[i];
		if (guard.isValid && (diff == 1 || diff == 0))
			Copy(inNumSamples, out, in);
		else
			Fill(inNumSamples, out, 0.f);
	}
```

The `(diff == 1 || diff == 0)` allows for both.

joshpar commented:
InFeedback is storing audio in its own buffer, and (I think?) always reads from that buffer, and writes to that buffer with what is in the audio bus currently. AudioControl relies on the pointer to audio samples that scsynth maintains (rather than an internal one), and I think that is where they difficulty is. We could add something from the RTAlloc for that... I'm just not sure how reliable it would be. Mostly - we need multiple test cases (really - unit tests) and probably much more time to get it to support all that.
Not sure why InFeedback doesn't glitch, unless it does but it just usually isn't noticeable. With the bug that AudioControl had, I'm guessing it was only heard because the same 64 samples KEPT getting looped (rather than being repeated just once).

joshpar commented:
(also - AudioControl can't take advantage of AudioBusGuard with its current set up - it may be something I can hack like the fix I put together this weekend, but it is messing with Graph and GraphDef a lot already ... and that makes me a little nervous)

telephon commented:
That seems to be be necessary anyhow (it is used in all other in ugens), something that might be necessary to make audio bus routing work on supernova.

Btw. if there is an issue about that sometimes you don't want the current feedback behavior, we could (once this works) add another osc message in a separate project. But for now, it is really good to slowly implement what is needed to make the current behavior work correctly.

As a first step, getting the feedback right might be a good next step.

telephon commented:
>I bet you'll see a buffer's worth of samples get repeated once.

hm, it would be good to actually produce this prediction.

Also I need to produce  a glitch with rearranging synth order. But when we discussed this, there were strong arguments for it to be unavoidable.

joshpar commented:
I'll put together a change in the pull-request to do what InFeedback is doing - and we can hear if there is a problem or not. Sound good? I can do that in a few hours.

jamshark70 commented:
> Also I need to produce a glitch with rearranging synth order.

Yes, I'd like to see that test case, because I haven't had the same experience. I've been arranging MixerChannels in order automatically for a good 10 years now, and I haven't noticed glitches. It's a key point of chucklib's design that you should be able to instantiate processes in the middle of a performance -- I've been doing this (creating mixers while other things are happening) for a long time, and I have never noticed glitches from the signal routing.

Granted, the rate of creating/destroying/reordering mixer channels is a lot slower than the rate for NodeProxies, so it's possible that I just haven't beaten it to death enough, but I have a feeling that starting with a good node order, and changing it to a different good node order, would be unlikely to glitch as long as the reordering happens at the same control block boundary (and the /n_order message should be able to handle something like 1600 IDs).

telephon commented:
Sure, adding new synths at runtime shouldn't be a problem. Do you also re-order synths that route into each other?

telephon commented:
… well in any case, removing feedbacks from NodeProxy would be a loss and not gain us much.

You should be able to express `Ndef(\y, { SinOsc.ar(200, Ndef.ar(\y) * MouseX.kr(0, 10)) * 0.1 }).play` also as:
```
Ndef(\y, { SinOsc.ar(200, \in.ar(0 ! 2) * MouseX.kr(0, 10)) * 0.1 }).play;
Ndef(\y) <<> Ndef(\y);
Ndef(\y, { SinOsc.ar(200, Ndef.ar(\y) * MouseX.kr(0, 10)) * 0.1 }).play;
```

This allows you not to worry about in what order you link one synth with another synth.

joshpar commented:
@telephon - I put in a similar case for checking for the last buffer, please take a look and see if it does what you think it should. Thanks!

joshpar commented:
any glitchiness that you hear? And - if you have a few minutes, please check with our other tests above (I can't at the moment, but I think they should be fine)

telephon commented:
I've checked all tests above, no glitches. This seems to fix the issue pretty well now. 

Other things could go into a new pull request (now or later), with `AudioBusGuard`, and maybe a second, non-feedbacking variant of `AudioControl`.

joshpar commented:
that rocks

\m/ *v* \m/

joshpar commented:
Let’s get this in for 3.9. I’ll look at the other stuff for a future improvement


/*
Josh Parmenter
www.realizedsound.net/josh
*/

> On Jul 31, 2017, at 15:32, Julian Rohrhuber <notifications@github.com> wrote:
> 
> I've checked all tests above, no glitches. This seems to fix the issue pretty well now.
> 
> Other things could go into a new pull request (now or later), with AudioBusGuard, and maybe a second, non-feedbacking variant of AudioControl.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

joshpar commented:
anyone else want to review or test these changes? @jamshark70 ?

brianlheim commented on server/plugins/IOUGens.cpp:
```diff
@@ -243,11 +263,21 @@ void AudioControl_next_1(AudioControl *unit, int inNumSamples)
-		case 2 :
-		    Copy(inNumSamples, out, *mapin);
+        case 2 : {
+            int thisChannelOffset = channelOffsets[unit->mSpecialIndex];
```
this block seems like a good place for a comment (also line 221 above). maybe consider also assigning `touched[thisChannelOffset]` to a temp variable so that the `if` below is easier to read.

brianlheim commented:
which of the comments above could I use as test cases?

joshpar commented:
at this point, I think I need to leave that question for @telephon - I don't fully understand everything going on with the Pbind examples myself, and I'm not sure which ones should make sound and which ones shouldn't.

jamshark70 commented:
> anyone else want to review or test these changes? @jamshark70 ?

I wouldn't mind, but it isn't going to happen until next week... got a show on Saturday and I'm not touching the environment until after that.

dkmayer commented:
As mentioned on users lists, examples from below now also work as expected !
Great job, thanks !

http://new-supercollider-mailing-lists-forums-use-these.2681727.n2.nabble.com/Pbind-audio-bus-mapping-issue-td7620326.html

joshpar commented:
Great! Thanks for checking. I’ll add comments and update formatting as per brian’s suggestion. New PR should be coming this evening.

> On Aug 2, 2017, at 2:43 PM, dkmayer <notifications@github.com> wrote:
> 
> As mentioned on users lists, examples from below now also work as expected !
> Great job, thanks !
> 
> http://new-supercollider-mailing-lists-forums-use-these.2681727.n2.nabble.com/Pbind-audio-bus-mapping-issue-td7620326.html <http://new-supercollider-mailing-lists-forums-use-these.2681727.n2.nabble.com/Pbind-audio-bus-mapping-issue-td7620326.html>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/supercollider/supercollider/pull/3063#issuecomment-319807660>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AADTrP28zjN1t3m5Au6A1S9jhZ5Z2ydDks5sUO2ZgaJpZM4OnMQZ>.
>

brianlheim commented on include/plugin_interface/SC_Graph.h:
```diff
@@ -35,6 +35,8 @@ struct Graph
 	uint32 mNumControls;
 	float *mControls;
 	float **mMapControls;
+    int32 *mAudioBusOffsets;
```
you missed one :P

vivid-synth commented:
Does this only fix for ```scsynth```, not ```supernova```?

brianlheim commented:
glad @vivid-synth asked - using the original testing code, this now crashes supernova (before, it just had the same issue).

```
(
SynthDef(\fxbus, { |out, in|
	var sig = InFeedback.ar(in, 2);
	ReplaceOut.ar(out, HPF.ar(sig, 800));
}).add;

SynthDef(\fx, { |out, a_in = #[0, 0]|
	ReplaceOut.ar(out, HPF.ar(a_in, 800));
}).add;

SynthDef(\boop, { |out, freq = 440, gate = 1, amp = 0.1|
	var sig = Mix(VarSaw.ar(freq * [1, 1.004])) * amp,
	eg = EnvGen.ar(Env.adsr(0.01, 0.1, 0.5, 0.001, curve: 0), gate, doneAction: 2);
	sig = LPF.ar(sig, 3000);
	Out.ar(out, (sig * eg).dup);
}).add;
)

s.boot;
s.scope;

b = Bus.audio(s, 2);

(
p = Pbind(
	\instrument, \boop,
	\degree, Pwhite(-7, 7, inf),
	\dur, 1,
	\legato, 0.25,
	\out, b
).play;
)

f = Synth.tail(s.defaultGroup, \fxbus, [in: b, out: 0]);  // OK

f.free;

(
s.makeBundle(nil, { // SUPERNOVA CRASH
	f = Synth.tail(s.defaultGroup, \fx);
	f.map(\a_in, b);
});
)

p.stop;
f.free;
```

brianlheim commented:
i'm getting a bad access here: https://github.com/supercollider/supercollider/blob/5f7d81162381370168e835cc9195f9344a581e50/server/plugins/IOUGens.cpp?utf8=%E2%9C%93#L232

brianlheim commented:
@joshpar , i took your changes from fdf1fb0debfdb42d5a05343f7b2563513f2203eb and applied them to similar spots in supernova. the original test code now works with supernova like it did with scsynth.

brianlheim commented on include/plugin_interface/SC_Graph.h:
```diff
@@ -35,6 +35,8 @@ struct Graph
 	float *mControls;
 	float **mMapControls;
+	int32 *mAudioBusOffsets;
+	uint32 mAudioBusOffset;
```
this appears to have been accidentally reintroduced - could you please remove it?

brianlheim replied:
or I can do it, I don't want this to get too confusing though :)

joshpar commented:
done

> On Aug 3, 2017, at 8:54 PM, Brian Heim <notifications@github.com> wrote:
> 
> @brianlheim commented on this pull request.
> 
> In include/plugin_interface/SC_Graph.h <https://github.com/supercollider/supercollider/pull/3063#discussion_r131307260>:
> 
> > @@ -35,6 +35,8 @@ struct Graph
>  	uint32 mNumControls;
>  	float *mControls;
>  	float **mMapControls;
> +	int32 *mAudioBusOffsets;
> +	uint32 mAudioBusOffset;
> or I can do it, I don't want this to get too confusing though :)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/supercollider/supercollider/pull/3063#discussion_r131307260>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AADTrAxtAxtyi5F9SesayMC63gXMjVdkks5sUpXpgaJpZM4OnMQZ>.
>

telephon commented:
great!

brianlheim commented:
all right. should any tests / other documentation be added for this? i am otherwise good to merge.

telephon commented:
i think it is ok to merge. we do need some systematic tests for mapping, but this is another issue

vivid-synth commented:
Thanks @joshpar and @brianlheim !

joshpar commented:
@snappizz - pretty sure this is ready to go

timblechmann commented on include/plugin_interface/SC_Graph.h:
```diff
@@ -35,6 +35,7 @@ struct Graph
 	uint32 mNumControls;
 	float *mControls;
 	float **mMapControls;
+	int32 *mAudioBusOffsets;
```
this looks like an ABI change. please increment https://github.com/supercollider/supercollider/blob/master/include/plugin_interface/SC_InterfaceTable.h#L24

vivid-synth replied:
@timblechmann changing an API from v1 to v2 gives me some pause -- will this potentially break existing code?

snappizz replied:
it doesn't break any code, but it does break binaries. you can't use server 3.9 with plugin binaries from 3.8, or vice versa.

dyfer commented:
I know this is closed, but it's causing a rather likely crashes in supernova: https://github.com/supercollider/supercollider/issues/3196 I think the referenced issue needs to be addressed before this gets included in the 3.9 release...

snappizz commented:
two issues that have since been identified as side effects of this PR: #3196 and #3266. i have marked them as required for 3.9.
